Add config flow to the Matrix integration

easyteacher opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This adds a config flow to the Matrix integration with no change in functions.

To validate the user input allows us to log in, the authentication part
of MatrixBot is separated into a new class named MatrixAuthentication.

Currently rooms and commands are not configurable through the config
flow because voluptuous_serialize does not support List type. 

_If converting a string to a list is acceptable, I will submit following changes later._

|Setup|Integration|Send Message|
|--|--|--|
|![图片](https://user-images.githubusercontent.com/23738961/148329778-341ddcbf-e645-49f5-aab8-f4dae99c9881.png)|![图片](https://user-images.githubusercontent.com/23738961/148329494-44ec6c24-3bea-4fc1-8a8d-0626cfcc658d.png)|![图片](https://user-images.githubusercontent.com/23738961/148330145-a9e1f709-8fe1-4d22-a5e2-81fd7624e92e.png)|




## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [x] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

easyteacher committed:
Move authentication part out of MatrixBot

To validate the user input allows us to log in, the authentication part
of MatrixBot is separated into a new class named MatrixAuthentication.

easyteacher committed:
Add config flow to the Matrix integration

This adds a config flow to the Matrix integration with no change in functions.

Currently rooms and commands are not configurable through the config
flow because voluptuous_serialize does not support List type.

easyteacher committed:
Add English translation to the Matrix integration

easyteacher committed:
Add config flow tests for the Matrix integration

easyteacher commented:
cc: @frenck

frenck commented on homeassistant/components/matrix/__init__.py:
```diff
@@ -52,17 +56,20 @@
-                vol.Optional(CONF_ROOMS, default=[]): vol.All(
-                    cv.ensure_list, [cv.string]
-                ),
+                vol.Optional(CONF_ROOMS, default=[]): ROOM_SCHEMA,
```
This change is not really needed or related, let's revert that.

frenck commented on homeassistant/components/matrix/__init__.py:
```diff
@@ -52,17 +56,20 @@
+    [cv.string],
+)
+
+CONFIG_SCHEMA: vol.Schema = vol.Schema(
```
This schema needs to be marked deprecated.

easyteacher replied:
Will do. But if it's marked deprecated, I will have to add `rooms` and `commands` settings to the options flow.

easyteacher replied:
```yaml
# Example configuration.yaml entry
matrix:
  homeserver: https://matrix.org
  username: "@my_matrix_user:matrix.org"
  password: supersecurepassword
  rooms:
    - "#hasstest:matrix.org"
  commands:
    - word: my_command
      name: my_command
```

How can I handle `rooms` and `commands` in the options flow? Is it possible to let the user type JSON strings? I see `ble_monitor` uses a more complex method to add/delete bluetooth devices, I am wondering if I will have to do so.

frenck commented on homeassistant/components/matrix/__init__.py:
```diff
@@ -82,9 +89,8 @@
 )
 
 
-def setup(hass: HomeAssistant, config: ConfigType) -> bool:
```
The `setup` method has been removed, but we would still need an integration setup, to allow triggering an import flow to import existing YAML configuration into the UI.

frenck commented on homeassistant/components/matrix/config_flow.py:
```diff
@@ -0,0 +1,131 @@
+        raise
+
+
+class ConfigFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
```
An import is missing: `async_step_import` to import existing YAML configuration.

frenck commented on homeassistant/components/matrix/config_flow.py:
```diff
@@ -0,0 +1,131 @@
+    ) -> FlowResult:
+        """Handle the initial step."""
+
+        # Only allow single instance because
+        if self._async_current_entries():
+            return self.async_abort(reason="already_configured")
```
Is there a specific reason to only accept a single instance?

easyteacher replied:
It should be possible to remove the limit, but if multiple instances are allowed, how can I name the registered service (matrix.send_message) for different instances?

frenck commented on homeassistant/components/matrix/config_flow.py:
```diff
@@ -0,0 +1,131 @@
+                await self.async_set_unique_id(info[CONF_ACCESS_TOKEN])
+                self._abort_if_unique_id_configured()
+
+                # CONF_ROOMS: set empty because we don't have any
+                user_input[CONF_ROOMS] = []
+
+                # CONF_COMMANDS: set empty because we don't have any
+                user_input[CONF_COMMANDS] = []
+
```
Why do we set these in the entry data?

easyteacher replied:
Will address after CONFIG_SCHEMA is marked deprecated.

frenck commented on homeassistant/components/matrix/config_flow.py:
```diff
@@ -0,0 +1,131 @@
+
+            except MatrixRequestError as ex:
+                # Error code definitions: https://spec.matrix.org/latest/client-server-api/#standard-error-response
+                error_code: Final[str] = json.loads(ex.content)["errcode"]
```
What is the json.loads fails?

frenck commented on homeassistant/components/matrix/config_flow.py:
```diff
@@ -0,0 +1,131 @@
+            CONF_ACCESS_TOKEN: token,
+        }
+
+    except (MatrixHttpLibError, MatrixRequestError):
+        # Re-raise so async_step_user can handle the exception
+        raise
```
Maybe not wrap it in a try/except block in that case?

frenck commented on tests/components/matrix/test_config_flow.py:
```diff
@@ -0,0 +1,166 @@
+FIXTURE_ACCESS_TOKEN: Final[str] = "syt_dummytoken_dummytoken_dummytoken"
+
+
+async def test_show_set_form(hass: HomeAssistant) -> None:
+    """Test that the setup form is served."""
+    result: FlowResult = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_USER}, data=None
+    )
+
+    assert result["type"] == RESULT_TYPE_FORM
+    assert result["step_id"] == "user"
```
Please complete the full configuration flow in this test.

frenck commented on tests/components/matrix/test_config_flow.py:
```diff
@@ -0,0 +1,166 @@
+
+        assert result["type"] == RESULT_TYPE_FORM
+        assert result["step_id"] == "user"
+        assert result["errors"] == {CONF_PASSWORD: "invalid_auth"}
```
Please complete the test; the user can recover from here.

frenck commented on tests/components/matrix/test_config_flow.py:
```diff
@@ -0,0 +1,166 @@
+
+        assert result["type"] == RESULT_TYPE_FORM
+        assert result["step_id"] == "user"
+        assert result["errors"] == {CONF_USERNAME: "invalid_auth"}
```
Please complete the test; the user can recover from here and complete the flow.

frenck commented on tests/components/matrix/test_config_flow.py:
```diff
@@ -0,0 +1,166 @@
+
+        assert result["type"] == RESULT_TYPE_FORM
+        assert result["step_id"] == "user"
+        assert result["errors"] == {CONF_BASE: "invalid_access_token"}
```
Please complete the test; the user can recover from here and complete the flow.

frenck commented on tests/components/matrix/test_config_flow.py:
```diff
@@ -0,0 +1,166 @@
+        assert result["errors"] == {CONF_BASE: "invalid_access_token"}
+
+
+async def test_login_success(hass: HomeAssistant) -> None:
```
This should be included in the first test

frenck reviewed (changes requested):
Some initial comments on a quick first review/pass.

easyteacher commented:
Thank you very much for the review. I will change the code asap. It seems the CI also gives some advice, and I will also take them.

easyteacher committed:
Add options flow to the Matrix integration

The options flow includes `rooms` and `commands` configuration entries.

easyteacher committed:
Add `set_listening_rooms` and `set_listening_commands` to MatrixBot

The two functions are required in update_listener.

1. `set_listening_rooms`: Set listening rooms and update the listeners.
2. `set_listening_commands`: Set listening commands.

easyteacher committed:
Add `async_step_import` and remove `setup`

Add `async_step_import` to import existing YAML configuration.

easyteacher commented:
Options flow is added so rooms and commands can be configured and functional like before. But test is not because it's a little complicated and I am still studying how to write the test.

![图片](https://user-images.githubusercontent.com/23738961/148688500-6a0ef880-d8e6-44f7-bf5c-c598ba8d9032.png)

![图片](https://user-images.githubusercontent.com/23738961/148688511-5eb6eec7-187c-466c-9bf2-cef440d74702.png)

easyteacher commented:
```
Traceback (most recent call last):
  File "/mnt/sdcard/linux/home/homeassistant/.homeassistant/deps/lib/python3.9/site-packages/matrix_client/client.py", line 490, in listen_forever
    self._sync(timeout_ms)
  File "/mnt/sdcard/linux/home/homeassistant/.homeassistant/deps/lib/python3.9/site-packages/matrix_client/client.py", line 624, in _sync
    room._put_event(event)
  File "/mnt/sdcard/linux/home/homeassistant/.homeassistant/deps/lib/python3.9/site-packages/matrix_client/room.py", line 311, in _put_event
    listener['callback'](self, event)
  File "/mnt/sdcard/linux/home/homeassistant/.local/lib/python3.9/site-packages/homeassistant/components/matrix/__init__.py", line 386, in _handle_room_message
    match = command[CONF_EXPRESSION].match(event["content"]["body"])
AttributeError: 'str' object has no attribute 'match'
```

Match by expression has some bugs.

easyteacher commented:
Another question: Is it possible to check the default values in the options config flow?

PaarthShah commented:
Hi @easyteacher! I don't know if you're still actively seeking to get this merged, but I definitely like the config flow implementation! I'm hoping to see later if it's already compatible/minimally conflicting with my PR here:
https://github.com/home-assistant/core/pull/72797

balloob commented:
This has been in draft for too long. Closing it. Feel free to open a new PR when you're ready to finish it.
